### PR TITLE
feat(provider/oraclebmcs): Network caching agent and provider

### PR DESCRIPTION
**Pull request by simonlord**
This is a PR for our first caching agent and provider pair for Networks with supporting classes and tests.

### CODE REVIEW

File: clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgent.groovy [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  OracleBMCSNetworkCachingAgent(String clouddriverUserAgentApplicationName,
+                                OracleBMCSNamedAccountCredentials credentials,
+                                ObjectMapper objectMapper) {
+    super(objectMapper, credentials, clouddriverUserAgentApplicationName)
+  }
+
+  @Override
+  CacheResult loadData(ProviderCache providerCache) {
+    List<Vcn> vcns = loadVcns()
+    buildCacheResult(vcns)
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgent.groovy]: Please add `return` statements to the methods that aren't just one-liners. (It's not a formal style requirement, but we're trying to be more consistent about it.)

File: clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgent.groovy [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      .compartmentId(credentials.compartmentId)
+      .build())
+    response.items
+  }
+
+  private CacheResult buildCacheResult(List<Vcn> vcns) {
+    log.info("Describing items in $agentType")
+
+    List<CacheData> data = vcns.collect { Vcn vcn ->
+      if(vcn.lifecycleState != Vcn.LifecycleState.Available){
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgent.groovy]: Missing space before and after `(`.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  @Shared
+  ObjectMapper objectMapper = new ObjectMapper()
+
+  def "agent has correct agentType"() {
+    setup:
+    def creds = Mock(OracleBMCSNamedAccountCredentials)
+    creds.name = "foo"
+    creds.compartmentId = "bar"
+    creds.region = Region.US_PHOENIX_1.regionId
+    def agent = new OracleBMCSNetworkCachingAgent("",creds, objectMapper)
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: Missing a space between `"",` and `creds`.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    agentType == expectedAgentType
+  }
+
+  def "agent handles null items in listNetworks rsp"(){
+    setup:
+    def creds = Mock(OracleBMCSNamedAccountCredentials)
+    def networkClient = Mock(VirtualNetworkClient)
+    networkClient.listVcns(_) >> ListVcnsResponse.builder().build()
+    creds.networkClient >> networkClient
+    def agent = new OracleBMCSNetworkCachingAgent("",creds, objectMapper)
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: Missing a space between "", and creds.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    def vcns = [
+      vcn,
+      new Vcn(null,null,null,null,null,"Another network","ocid.vcn.321",Vcn.LifecycleState.Terminated,null),
+      new Vcn(null,null,null,null,null,"Yet Another network","ocid.vcn.531",Vcn.LifecycleState.Terminating,null),
+      new Vcn(null,null,null,null,null,"Coming soon network","ocid.vcn.878",Vcn.LifecycleState.Provisioning,null)
+    ]
+
+    networkClient.listVcns(_) >> ListVcnsResponse.builder().items(vcns).build()
+    creds.networkClient >> networkClient
+    def agent = new OracleBMCSNetworkCachingAgent("",creds, objectMapper)
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: Missing a space between "", and creds.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    networkClient.listVcns(_) >> ListVcnsResponse.builder().build()
+    creds.networkClient >> networkClient
+    def agent = new OracleBMCSNetworkCachingAgent("",creds, objectMapper)
+
+    when:
+    def cacheResult = agent.loadData(null)
+
+    then:
+    cacheResult != null
+    cacheResult.cacheResults.containsKey((Keys.Namespace.NETWORKS.ns))
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: I don't think the two sets of parens are needed here.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    networkClient.listVcns(_) >> ListVcnsResponse.builder().items(vcns).build()
+    creds.networkClient >> networkClient
+    def agent = new OracleBMCSNetworkCachingAgent("",creds, objectMapper)
+
+    when:
+    def cacheResult = agent.loadData(null)
+
+    then:
+    cacheResult != null
+    cacheResult.cacheResults.containsKey((Keys.Namespace.NETWORKS.ns))
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: Or for these either.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy (hunk 6 above)
Comment (simonlord) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: Done - not sure why github hasn't outdated this chunk

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy (hunk 7 above)
Comment (simonlord) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: Done - not sure why github hasn't outdated this chunk

**Comment by simonlord** [comment]
Ok all style changes made :)

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy (hunk 6 above)
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSNetworkCachingAgentSpec.groovy]: That looked like some propagation delay. It just now went away.

**Pull request merged**